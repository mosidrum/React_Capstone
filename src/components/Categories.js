import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { allCountries, getCategoryOfCountries, isLoading } from '../redux/categorySlice';
import '../css/categories.css';
import loadingImage from '../assets/loading_3.gif';

const Categories = () => {
  const allCategories = useSelector(allCountries);
  const loading = useSelector(isLoading);
  const dispatch = useDispatch();
  const cat = allCategories.map((each) => (each.region));
  const categories = [...new Set(cat)];
  console.log(categories);

  const handleSubmit = (id) => {
    dispatch(getCategoryOfCountries(id));
  };

  const images = [
    'https://i.ibb.co/zZ16v9B/Asia.png',
    'https://i.ibb.co/3fLsYTF/Europe.png',
    'https://i.ibb.co/9tHqvcc/Africa.png',
    'https://i.ibb.co/ZN1sYPg/Oceania.png',
    'https://i.ibb.co/yNb2x44/Americas.png',
    'https://i.ibb.co/7rNfzyr/Polar.png',
    'https://i.ibb.co/mJ7TR65/Antarctic.png',
    'https://i.ibb.co/mJ7TR65/Antarctic.png',
  ];

  const categoryImage = categories.reduce((imageMaped, category, index) => {
    imageMaped[category] = images[index];
    return imageMaped;
  }, {});

  const categoryCounts = categories.reduce((counts, category) => {
    const count = allCategories.filter((item) => item.region === category).length;
    counts[category] = count;
    return counts;
  }, {});

  return (
    <div className="categories">
      {loading && (
        <img
          src={loadingImage}
          alt="loading gif"
          className="loading"
        />
      )}
      {!loading && categories.map((category) => (
        <div
          type="submit"
          className="category"
          key={category}
        >
          <img src={categoryImage[category]} alt={category} />
          <NavLink
            to="/countries"
          >
            <button
              className="cat"
              onClick={() => handleSubmit(category)}
              type="submit"
            >
              {category}
              (
              {categoryCounts[category]}
              country
              )
            </button>
          </NavLink>
        </div>
      ))}
    </div>
  );
};

export default Categories;
