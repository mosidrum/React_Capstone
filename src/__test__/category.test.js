import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { useDispatch, useSelector } from 'react-redux';
import Categories from '../components/Categories';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('Categories component', () => {
  it('should render categories correctly', () => {
    const allCategories = [
      { region: 'Asia' },
      { region: 'Europe' },
      { region: 'Africa' },
    ];
    const isLoading = false;

    useSelector.mockReturnValueOnce(allCategories);
    useSelector.mockReturnValueOnce(isLoading);
    useDispatch.mockReturnValue(jest.fn());

    const { container } = render(
      <MemoryRouter>
        <Categories />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});