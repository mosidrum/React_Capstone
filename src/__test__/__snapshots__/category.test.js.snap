// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories component should render categories correctly 1`] = `
<div>
  <div
    class="categories"
  >
    <div
      class="category"
      type="submit"
    >
      <img
        alt="Asia"
        src="https://i.ibb.co/zZ16v9B/Asia.png"
      />
      <a
        class=""
        href="/countries"
      >
        <button
          class="cat"
          type="submit"
        >
          Asia
          (
          1
          country )
        </button>
      </a>
    </div>
    <div
      class="category"
      type="submit"
    >
      <img
        alt="Europe"
        src="https://i.ibb.co/3fLsYTF/Europe.png"
      />
      <a
        class=""
        href="/countries"
      >
        <button
          class="cat"
          type="submit"
        >
          Europe
          (
          1
          country )
        </button>
      </a>
    </div>
    <div
      class="category"
      type="submit"
    >
      <img
        alt="Africa"
        src="https://i.ibb.co/9tHqvcc/Africa.png"
      />
      <a
        class=""
        href="/countries"
      >
        <button
          class="cat"
          type="submit"
        >
          Africa
          (
          1
          country )
        </button>
      </a>
    </div>
  </div>
</div>
`;
